src/models/utils.ts

```diff
@@ -0,0 +1,30 @@
+import * as _ from 'lodash';
+
+export interface WithId {
+	id: number;
+}
+
+export interface PineResourceBase extends WithId {
+	created_at: Date;
+	modified_at: Date | null;
+}
+
+export interface PineDeferred {
+	__id: number;
+}
+
+/**
+ * When not selected-out holds a deferred.
+ * When expanded hold an array with a single element.
+ */
+export type NavigationResource<T = WithId> = T[] | PineDeferred;
+
+/**
+ * When expanded holds an array, otherwise the property is not present.
+ * Selecting is not suggested,
+ * in that case it holds a deferred to the original resource.
+ */
+export type ReverseNavigationResource<T = WithId> = T[] | undefined;
+
+export const getExpanded = <T>(obj: NavigationResource<T>) =>
```

Comment:
I think if you use
```suggestion
export function getExpanded <T>(obj: T[]): T | undefined
export function getExpanded (obj: NavigationResource<T>): undefined
export function getExpanded <T>(obj: NavigationResource<T>): T | undefined {
```
then you get a more accurate typing and it can catch cases where you're passing in a PineDeferred and it's impossible for the result to ever be anything other than `undefined`, eg
```ts
const x = getExpanded(pineDeferred);
if (x != null) {
	// typeof x === never
}
```
Unfortunately we cannot rely on `<T>(obj: T[]): T` because we might pass in `[]` which would still return undefined

Reply:
I think that what you suggest is done by:
```ts
export function getExpanded <T>(obj: PineDeferred): undefined;
export function getExpanded <T>(obj: NavigationResource<T>): T | undefined;
export function getExpanded <T>(obj: NavigationResource<T>): T | undefined {
	return (_.isArray(obj) && obj[0]) || undefined;
}
```